compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirCallableMemberDeclaration
+import org.jetbrains.kotlin.fir.declarations.FirRegularClass
+import org.jetbrains.kotlin.fir.declarations.modality
+
+object FirOpenMemberChecker : FirRegularClassChecker() {
+    override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration.canHaveOpenMembers()) return
+        for (memberDeclaration in declaration.declarations) {
+            if (memberDeclaration !is FirCallableMemberDeclaration<*>) continue
+            val source = memberDeclaration.source
```

Comment:
One more thing to check here. We don't want to report on _fake_ source, either.

Reply:
If we don't report fake, the following `x` will be missed

```
value class A8(open val x: Int)
```

Reply:
Yeah, as discussed offline, let's allow fake source for value parameter of `inline`/`value` classes.

Reply:
But, due to the modifier list from source (below), it's better to filter out some fake sources here.